text/0000-manually-drop.md

```diff
@@ -0,0 +1,264 @@
+  you avoid adding an (unnecessary) additional work for the
+  discriminant?), and especially not if we ever get generic number
+  parameters, so something like `struct SmallVec<T, n> { ... inner: ManuallyDrop<[T, .. n]> }`
+  would work, allowing for `SmallVec<T, 2>`, `SmallVec<T, 100>` etc.
+
+- A struct `UninterpretedBytesOfSize<T>` equal to
+  `[u8, .. size_of::<T>()]`, that is, a chunk of memory large enough
+  to store a `T`, but treated as raw memory (i.e. `u8`s). This has the
+  (large) downside of losing all type information, interfering with
+  the compiler's reachability analysis (e.g. for `UnsafeCell`), and
+  making it easier for the programmer to make mistakes w.r.t. an
+  incorrect or forgotten coercion (it's would be identical to C's
+  `void*`).  This is getting more feasible with [`const fn`]
+  support. There needs to be support for explicitly specifying the
+  alignment of the type too (to match `T`).
+
+- Change drop to have (semantically) take full ownership of its
+  contents, so that `mem::forget` works, e.g. `trait Drop { fn
+  drop(self); }` or a design like that in [@eddyb's comment].
+
+- Make no change and just perform manual ownership control via
+  `Option<T>` (or something similar, designed to avoid invalid layout
+  optimisations): a data type can store `Option<T>` instead of `T`
+  with the invariant that the `Option` is always `Some` except for
+  when the destructor runs. This allows one to implement the previous
+  alternative with `take`. E.g. `ManuallyDrop` can be shimmed as
+  something like:
+
+  ```rust
+  struct ManuallyDrop<T> { x: Option<T> }
+  impl<T> ManuallyDrop<T> {
+      const fn new(x: T) -> ManuallyDrop<T> {
+          ManuallyDrop { x: Some(x) }
+      }
+      fn get(&self) -> *const T {
+          self.x.as_ref().unwrap()
+      }
+      fn get_mut(&mut self) -> *mut T {
+          self.x.as_ref_mut().unwrap()
+      }
+      unsafe fn into_inner(self) -> T {
+          self.x.unwrap()
+      }
+  }
+
+  impl Drop for ManuallyDrop {
+      fn drop(&mut self) { mem::forget(self.x.take()) }
+  }
+  ```
+
+  This approach is used by [arrayvec] but has many downsides:
+
+  - (much) larger compiled code when not using `unsafe` (`unwrap`
+    includes branches and unwinding etc.)
+  - more invariants to track
+  - larger data types
+
+[`const fn`]: https://github.com/rust-lang/rfcs/blob/master/text/0911-const-fn.md
+[@eddyb's comment]: https://github.com/rust-lang/rfcs/pull/197#issuecomment-110850383
+
+# Unresolved questions
+
+- Should a `ManuallyDrop<T>` always be `Copy`? It no longer has a
+  destructor and so the only risk of double freeing (etc.) would be
+  when the user writes such a thing. This would allow it to be a
+  maximally flexible building-block, but I cannot think of a specific
+  use-case for `ManuallyDrop<NonCopy>` to be `Copy`. Being `Copy`
+  would make `ManuallyDrop` entirely the inline equivalent of `*const`
+  and `*mut`, since they are both `Copy` always.
```

Comment:
This might be useful for RawVec to avoid having to do this:

```rust
        unsafe {
            // need to use ptr::read to unsafely move the buf out since it's
            // not Copy.
            let buf = ptr::read(&self.buf);
            let len = self.len;
            mem::forget(self);

            IntoIter {
                start: *buf.ptr,
                end: buf.ptr.offset(len as isize),
                _buf: buf,
            }
        }
```

But the code is already so blatantly unsafe that it's kinda "whatever"

(From http://cglab.ca/~abeinges/blah/turpl/_book/vec.html )